packages/api/lib/ems.js

```diff
@@ -0,0 +1,291 @@
+'use strict';
+
+const fs = require('fs');
+const moment = require('moment');
+const os = require('os');
+const path = require('path');
+const aws = require('@cumulus/common/aws');
+const log = require('@cumulus/common/log');
+
+const { deconstructCollectionId } = require('../es/indexer');
+const { Search, defaultIndexAlias } = require('../es/search');
+
+/**
+ * This module provides functionalities to generate EMS reports.
+ * The following environment variables are used:
+ * process.env.ES_SCROLL_SIZE: default to defaultESScrollSize
+ * process.env.ES_INDEX: set for testing purpose, default to defaultIndexAlias
+ * process.env.ems_provider: default to 'cumulus'
+ * process.env.stackName: it's used as part of the report filename
+ * process.env.bucket: the bucket to store the generated reports
+ */
+
+const defaultESScrollSize = 1000;
+
+/**
+ * For each EMS report type (ingest, archive, delete),
+ * map the EMS fields to CUMULUS granule/deletedgranule record fields,
+ */
+const emsMappings = {
+  ingest: {
+    dbID: 'granuleId',
```

Comment:
You could probably save a ton of code elsewhere and be more clear by having the mapping map from a key to either a string (property) or a function, i.e.

    const fieldToDate = function(field) {
      return (granule) => someDateConversion(granule[field]);
    };

    const emsMappings = {
      dbId: 'granuleId',
      processingStartDateTime: fieldToDate('processingStartDate')
    };

Reply:
I see what you meant.  But there are quite a few conversion cases, and we either use function for all or none.  I guess I choose none